cpp/ql/src/experimental/Security/CWE/CWE-788/AccessOfMemoryLocationAfterEndOfBufferUsingStrncat.ql

```diff
@@ -0,0 +1,65 @@
+/**
+ * @name Access Of Memory Location After The End Of A Buffer Using Strncat
+ * @description --Calls of the form strncat(dest, src, sizeof (dest) - strlen (dest)) set the third argument to one more than possible.
+ *              --So when the buffer is full, the expression sizeof(dest) - strlen (dest) will be equal to one, and not zero as the programmer might think.
+ *              --Making a call of this type may result in a zero byte being written just outside the buffer.
+ * @kind problem
+ * @id cpp/access-memory-location-after-end-buffer
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-788
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.HashCons
+
+/**
+ * A call to `strncat` of the form `strncat(buff, str, someExpr - strlen(buf))`, for some expression `someExpr` equal to `sizeof(buff)`.
+ */
+class WrongCallStrncat extends FunctionCall {
+  Expr leftsomeExpr;
+
+  WrongCallStrncat() {
+    this.getTarget().hasGlobalOrStdName("strncat") and
+    // the expression of the first argument in `strncat` and `strnlen` is identical
+    hashCons(this.getArgument(0)) =
```

Comment:
I think what you really want to use here is global value numbering (from `semmle.code.cpp.valuenumbering.GlobalValueNumbering`) instead of hash consing. `globalValueNumber(e1) = globalValueNumber(e2)` guarentees that the expressions `e1` and `e2` evaluate to the same value, where's `hashCons(e1) = hashCons(e2)` holds if `e1` and `e2` has "identical syntax", but not necessarily evaluate to the same value.

So for instance, if you use global value numbering you'll flag a call such as:
```cpp
char* dest_copy = dest;
strncat(dest, src, sizeof(dest) - strlen(dest_copy));
```

which you won't flag using hash consing.